pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -321,6 +323,8 @@ public void addComplete(Position position, ByteBuf entryData, Object ctx) {
                         }
                         txnCommittedCounter.increment();
                         completableFuture.complete(null);
+                        // need to update the last data message published timestamp
+                        topic.setLastDataMessagePublishedTimestamp(Clock.systemUTC().millis());
```

Comment:
better to move to https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L3907

Reply:
LastDataMessagePublishedTimestamp is used to determined whether there are new messages available for reading. A transactional production should not update it as the uncommitted messages are not visible.

Reply:
I've checked the source code, it is for `ReplicatedSubscriptionsController`, use `LastDataMessagePublishedTimestamp` to determine if we need to snapshot. In an ideal situation, we need to update `LastDataMessagePublishedTimestamp` after txn committed. But topic transaction buffer just maintain the first message position of the txn. Maybe we can update `LastDataMessagePublishedTimestamp` after txn committed.

Reply:
And maybe we need consider race condition while updating it.

Reply:
Actually, I think update `LastDataMessagePublishedTimestamp` before or after the txn committed doesn't matter.
Because if there is a non-txn message published to the topic before some txn committed will also update it.

Reply:
I think we need to update it after the txn is aborted or commited, both of which can move forward the max read position.
It is normal that we produce transactional messages only for a specific topic, in such case we don't update `LastDataMessagePublishedTimestamp` at all for now.